### ISSUE TITLE
Allow release mode builds to run on simulators

### ISSUE THREAD

**Issue by gavindoughtie** [flutter/flutter#11754]
**Original issue**: Current documentation (https://github.com/flutter/flutter/wiki/Flutter%27s-modes) states that Flutter release mode builds cannot be run on the Simulator. For many applications, the iOS Simulators are sufficiently high fidelity to do 99% of development work without going through the additional friction of deploying to devices regularly. It's unfortunate that, unlike in native application development, you can't get an accurate understanding of your release mode behaviors with Flutter.

**Merging #190883** by @jeffwall-curlewlabs: 

### Use case

*(Observed on Flutter 3.44.8, stable.)*

I need to run `integration_test` against an **optimized** (AOT-compiled, tree-shaken) build of my app on the iOS Simulator, to catch release-only *correctness* regressions before they ship.  Today that is impossible: the tool refuses both `--profile` and `--release` for any simulator, so debug/JIT is the only mode a simulator can run.

**This is a platform-parity gap, not a new capability.** The same coverage already works on Android with no special setup: `flutter drive --profile` against an emulator produces an AOT, tree-shaken build - including an x86_64 `libapp.so` slice for the emulator - and runs the integration test against it.   iOS has no equivalent at any optimization level, on any simulator.

**The failure class this catches.** My app compiles a test-only sign-in path in behind a `bool.fromEnvironment` const, so in store builds that path is a dead branch the tree-shaker removes by design.  That is exactly the shape where an optimizer-only break hides: a branch that gets dropped, a plugin registration that gets shaken out, a codegen or reflection path that only resolves under JIT.  Debug builds skip all of it, so every automated test I can run on a simulator is blind to that entire class.  Concretely: I can catch a tree-shaking regression on an Android emulator today, and I cannot catch the same regression on iOS without a device.

**This is explicitly not a performance request**, which I think is where the existing discussion has settled.  #11754, #58988 and #149420 all end up weighing whether profile-mode *measurements* on a simulator are meaningful, and the answer given there - software rendering, so the numbers don't represent real hardware - is correct.  I'm not disputing it and I'm not asking you to revisit it.  I don't want timings.  I want to execute AOT-compiled code and assert on behavior.  Rendering fidelity is irrelevant to that: I'm asserting that a code path still exists and still runs, not how fast it draws.

**Alternatives considered:**

- **A physical iOS device in CI.** This works, and it's the only route available today.  It means buying, provisioning, signing and babysitting hardware wired to a macOS runner in order to get coverage that costs nothing on Android.  It's a workaround, not a solution.
- **`--profile` instead of `--release`.** Refused by the same rule.  On iOS, profile isn't the cheaper fallback it is on Android.
- **A custom local engine.** Self-defeating: the point is to test the toolchain that actually ships, and this trades that away.  (#38933 also suggests the local-engine path for iOS profile/release isn't currently healthy.)
- **A package on pub.dev.** Not addressable from package space - this is build mode and artifact selection inside the tool and engine.

### Proposal

Make it possible to build and run an AOT-compiled Flutter app on the iOS Simulator, so `integration_test` / `flutter drive` can execute optimized code there.  I have no stake in which mode is the vehicle - profile alone would fully serve this for my needs.

Some observations from the shipped 3.44.8 artifacts - please correct anything I've misread:

- **The refusal is mode-shaped rather than capability-shaped.**  `isEmulatorBuildMode()` in `build_info.dart` returns true only for `BuildMode.debug`, and `build_ios.dart` throws off it.  There's a second, independent refusal inside the AOT target itself (`build_system/targets/ios.dart`) when `SDKROOT` is a simulator SDK.  #58886 previously framed the first of these as better expressed in terms of supported AOT architectures; that framing still looks right to me, though clearly it isn't sufficient on its own given the second gate and the artifacts below.
- **The `ios-profile` and `ios-release` artifact sets do ship simulator slices** in `Flutter.xcframework`, which is what initially made me think the pieces were already present.  Comparing them against the device slices suggests they're JIT-shaped: the simulator slices in all three sets carry the VM-service and Dart front-end symbols that the device profile/release slices don't.  (Method, in case it's useful or in case I've got this wrong: `strings` over each `Flutter.framework/Flutter`, grepping for `vm_service` and `kernel-service`.)
- **There appears to be no simulator-targeted AOT compiler.**  Each iOS artifact directory ships a single device `gen_snapshot_arm64`, and its Mach-O options (`--macho-min-os-version`, `--macho-install-name`, `--macho-rpath`, and friends) don't seem to include a platform selector - so I don't think there's currently a way to emit a snapshot stamped `PLATFORM_IOSSIMULATOR`.

I would be OK with:
- **Profile only.**  Release buys nothing here over profile.  Profile is AOT and tree-shaken, which is the entire ask.
- **arm64 simulator only.**  Apple-silicon hosts only.  x86_64 simulator support would be nice to have, and I'd guess it's the more expensive half.
- **Opt-in artifact.**  Gated behind a `flutter precache` flag or a config setting, so the bytes aren't in every developer's default download.

Any one of those would give iOS the same release-correctness coverage Android emulators already provide, which is all I'm after.

**Comment by Hixie** [comment]
The release mode behaviours should be the same in debug and release mode, except for performance (for which the simulator doesn't give you an accurate understanding anyway).

**Comment by Hixie** [comment]
@collinjackson mentions that you might want to test other things about release builds, e.g. crash reporting.

We don't plan to support this in the near term, but I'm leaving this bug open to track the topic.

**Comment by AppleEducate** [comment]
This is also important to be able to capture screenshots with Fastlane in release mode on simulators.

**Comment by MisterJimson** [comment]
I also don't see a reason to restrict release mode builds to physical devices. Developers may need to test release mode builds but not have immediate access to a physical device.

**Comment by AppleEducate** [comment]
Especially for developers that are on Mac and may not have an Android Phone.

Rody Davis Jr.
Full Stack Mobile Developer
Github | Website | Appstore | LinkedIn
Twitter | Facebook | YouTube

**Comment by Xe** [comment]
This bit me when I was trying to get iOS screenshots for devices I do not own (like the iPhone 8 Plus).

**Comment by MisterJimson** [comment]
Is there any other cross platform framework that has this restriction? I know Xamarin does not.

**Comment by Hixie** [comment]
I guess we could replace the outright ban with a prominent warning. If anyone wants to implement that, we would definitely consider a PR implementing it.

**Comment by csga5000** [comment]
I second that this is needed for screenshots.  In the mean time is there any way to get rid of the debug banner?  Because I don't have all the devices I need screenshots for.

**Comment by csga5000** [comment]
https://stackoverflow.com/questions/48893935/how-to-remove-debug-banner-in-flutter-on-android-emulator

**Comment by jamesderlin** [comment]
> The release mode behaviours should be the same in debug and release mode, except for performance (for which the simulator doesn't give you an accurate understanding anyway).

Behaviors for *applications* should be the same.  However, I think that the ability to run release builds in simulators would be a big help to testing changes to the Flutter engine because the engine does very different things in the different modes.

**Comment by matpag** [comment]
Sometimes it happens that for fast-prototyping I don't have an apple enrolled developer account on the machine.

Without the ability to test a release build on the simulator I can't even see how much space will take the release bundle.
I'm able to do this both on Xamarin.Forms and React Native, don't see why we can't do it in Flutter.

Please make this command `flutter run --release -d <ios-simulator-id>` work, thank you :)

**Comment by TigPT** [comment]
+1

**Comment by Axort** [comment]
It's possible to remove the debug banner by setting the debugShowChekedModeBanner MaterialApp property to false, e.g.

`return new MaterialApp(title: 'MyApp', debugShowCheckedModeBanner: false,...);`

**Comment by vitor-gyant** [comment]
@Hixie Is there anyway to have flutter compiling in release mode for the Simulator arch? Thanks

**Comment by Hixie** [comment]
What is your use case?

**Comment by vitor-gyant** [comment]
@Hixie

My use case is that I've to redistribute a part of our flutter app as a native SDK. 

To do that, I've basically created a thin wrapper that implements the flutter app2app plumbing to avoid exposing the flutter APIs to our client and publish Flutter.framework and App.framework as private pods on our private repo. My SDK podspec declares App and Flutter as dependencies and all the plugins are compiled as static libraries. 

However, for development purposes I need to create fat frameworks with device and simulator archs. Normally we would compile all of them in release. However, it seems that release mode was explicitly blocked on the xcode_backend.sh script. 

Maybe you could clarify the limitations for such restriction or if it would be possible to compile it in release mode somehow. 

Moreover, the i386 arch requires us to change flutter gn to allow it to be compiled in a x86_64 machine.

Thanks in advance,

**Comment by Hixie** [comment]
Interesting use case, definitely not something we've considered before...

The reason we haven't allowed it so far is that people tend to use it to test performance, but then they get disappointed when the performance on a simulator really has no bearing on real device performance.

**Comment by vitor-gyant** [comment]
I see. But Apple also has the some problems and they basically make a good statement on the fact that performance must be measured on the device. Would you guys consider to remove this restriction ? Is there any patch on my side I could do to compile for release ? Thanks

**Comment by vitor-gyant** [comment]
@Hixie Sorry for pushing this. Is there a way to compile flutter locally for x86_64 and i386 in release ?

**Comment by vitor-gyant** [comment]
Any update here ? Glad to help if someone would provide some highlights.

**Comment by anondev32** [comment]
This feature is also important for developers who work remotely and want to present their current version / work via ScreenSharing to their clients for example.

**Comment by lukasschneiderapps** [comment]
> Interesting use case, definitely not something we've considered before...
> 
> The reason we haven't allowed it so far is that people tend to use it to test performance, but then they get disappointed when the performance on a simulator really has no bearing on real device performance.

For me the reason would be testing my Flutter app in release mode without owning a physical IOS device.

**Comment by vitor-gyant** [comment]
Now that bitcode support is out for testing this is the remaining feature to ship iOS native frameworks with flutter. Any idea if this could be a short term priority ? Thanks.

**Comment by vitor-gyant** [comment]
@xster @jonahwilliams  @eseidelGoogle Is this something that was scheduled for a near future ? Is there anything the community could help with ? Thanks.

**Comment by eseidelGoogle** [comment]
The work needing done is mostly on the engine side I suspect.  Similar to what @jonahwilliams did recently for adding support for release x86.  See PRs linked from https://github.com/flutter/flutter/issues/9253.

**Comment by GioPan04** [comment]
Any updates?

**Comment by sma** [comment]
We distribute a Flutter application as an iOS framework (to hide implementation details) and this is currently quite painful because the user must use both Debug and Release builds of the frameworks because we cannot simply distribute the Release build and tell the customer to only use a real device for development. Therefore, we really would like to have a Release version iOS framework build that runs on the Simulator.

**Comment by joaoarmando** [comment]
The mainly reason of what i need it is to make a video screen record. The debug version is too slow. When i use the release version "flutter build apk" on android it's look pretty great! The app is so smoothly but on ios simulator by using the debug version is too poor.

**Comment by nbrownus** [comment]
It is also not just the banner. If you have code that branches on `kDebugMode`, `kReleaseMode`, or `kProfileMode`, you can not run on a simulator and have the correct experience.

**Comment by jpiabrantes** [comment]
> The release mode behaviours should be the same in debug and release mode, except for performance (for which the simulator doesn't give you an accurate understanding anyway).

It should, but sometimes they are not. I would like to know when they are not the same fast, using the simulator.

**Comment by britannio** [comment]
I'd like to use the IOS simulator to record parts of my app to create [App Previews](https://developer.apple.com/app-store/app-previews/) for my App Store listing. I don't have access to a 6.5" Iphone so the closest alternative is the IOS simulator. I don't expect performance of the simulator to be representative of real world app performance but my assumption is that it'd be suitable enough to record a few videos for this purpose.

**Comment by huainanzi** [comment]
I use flutter build ios-framework build Release And Debug frameWork. I try use lipo -create Debug/app.framework Release/app.framework. I Can Run IOS simulator but I get error and blank "[VERBOSE-2:engine.cc(134)] Engine run configuration was invalid."

**Comment by murtuzasaify** [comment]
Is there a solution for this ? As there are quit a times developer needs to validate things for release configuration and  its bit frustrating to not able to run it on simulator.

**Comment by 1ucas** [comment]
Hello everybody.
The year is 2021 and we're now joining the other devs suffering from this limitation imposed by the release version of Flutter.
Here it is our use case: 

We have a multi module - 100% native - app. All modules are integrated as a dependency of the main app, via Cocoapods (private pods). Until now, everything works just fine. And on our CI, among other validations, the app goes through the current Steps: 

1. Download all dependencies (pod install)
2. build
3. unitTests
4. UI + Snapshot Tests
5. Archive 
6. Deploy to Testflight

**The situation:**
If `Steps 3 and 4 target iOS simulators`, so we can get the DIFFs for different devices and check how our app is going to look on different screens. On the other hand, for the sake of performance, `Steps 5 and 6 must target only the ARM architecture` (generic iOS device).

**The problem:**
Ever since we added Flutter as a dependency, it totally broke the current pipeline.
The main issue is that to make `Steps 3 and 4` run, we have to use the **Debug** versions of the .xcframework generated. But the `Steps 5 and 6` must use the **Release** variant. 

**The solutions:**
To make it all work, we have to re-run `pod install` between `Steps 4 and 5` to change the reference of the files we are pointing at. Or manually edit the app project file. Either solutions, terrible, slow and error-prone.
What we believe it would be the most appropriate way: **Run all Steps already referencing the Release version of the .xcframework**

Anyone got any tips on how to proceed on this situation?

**Comment by nstrelow** [comment]
Is there any interest in pursuing this and adding support?

**Comment by 1ucas** [comment]
Sure!

**Comment by ristiisa** [comment]
> Is there any interest in pursuing this and adding support?

yes

**Comment by ashish-d-hh** [comment]
Any update for running release build on iOS Simulators ?

**Comment by gustavogervasio** [comment]
Any update for running release build on iOS Simulators ?

**Comment by mlb5000** [comment]
Currently in the same situation as @sma. It makes it very difficult for our clients to integrate with a simple CocoaPod workflow.

**Comment by sostenesgomes** [comment]
Any updates?

**Comment by thilojaeggi** [comment]
Any updates or at least a way to hide layout overflow errors, I just have one which ruins the screenshot.

**Comment by GeylanKalafMohe** [comment]
Any update please?? This is a basic feature which should be supported. How can we test if the release is working on an iPhone 15 without the physical device regardless of debug mode?

**Comment by justoverclockl** [comment]
any news?

**Comment by filippo-veneri** [comment]
It's 2024 and this issue is still getting ignored, after **7** years. Any workaround?

**Comment by CristovoXDGM** [comment]
Why keep ignoring everyone, this feature would help a lot, what is the meaning of not add this command to run on ios ? Android have this, why not on ios?

**Comment by CristovoXDGM** [comment]
currently i ran into a bug that only happens in release mode, why not allow us to test  in release mode ?

**Comment by cl3m** [comment]
@Hixie This limitation is quite cumbersome for embedding a flutter module in release mode in another app and test on the simulator/device on iOS. Could you reconsider ? How are is this change ? Any pointer for a workaround/PR ?

**Comment by Hixie** [comment]
This is not something anyone who contributes to Flutter regularly has [said is on their radar for 2024](https://github.com/flutter/flutter/wiki/Roadmap), but we welcome contributions from everyone, so if anyone is interested in implementing this, please do feel free to do so. The main thing is we would want a prominent warning that performance in a simulator is never representative, even in release builds. If you need help getting started, check our [contributing guide](https://github.com/flutter/flutter/blob/main/CONTRIBUTING.md), join the Discord (link in the guide) and ask for help in the #hackers-new channel.

**Comment by cl3m** [comment]
Release aar works on the Android emulator without warning so I'm not sure iOS simulator needs additional warning for the embedding use case. I found a workaround for my use case by the way: 

1. Build the framework using `flutter build ios-framework`
2. Replace the ios-arm64_x86_64-simulator part of the App.xcframework release build with the one from the debug build.
3. Use the Flutter.xcframework from the release build
4. Run on the iOS simulator and device.

**Comment by davidstearns911** [comment]
Any update on this? Causing major issues.

**Comment by krll-kov** [comment]
Any progress?

**Comment by chunhtai** [comment]
This will require config change in xcode project

**Comment by tvolkert** [comment]
FYI, as of December 2024, this is still not something we plan to work on, but we welcome contributions if anyone is interested in implementing this.  See https://github.com/flutter/flutter/issues/11754#issuecomment-1960436803

**Comment by erfanegtfi** [comment]
It seems this problem is not going to be solved even after about six years.

**Comment by tvolkert** [comment]
@erfanegtfi see my last comment

**Comment by FelixMittermeier** [comment]
I got it working on Apple Silicon iOS simulators and created a PR for it: https://github.com/flutter/flutter/pull/191002
However as I don't have access to an Intel based Mac I am not able to test if it correctly works on Intel iOS simulators as well. Would be great if anyone with access to one could try it out and give feedback there. Thanks

**Comment by cbracken** [comment]
@FelixMittermeier Thanks for putting together a proof-of concept. Before we go further, questions we probably want to address are:

* What's our support model for this look like (supported, best effort, not supported)? @jmagman @vashworth 
* Do we have infra capacity for the extra builds? @jtmcdole 

Probably a bunch more.

**Comment by jtmcdole** [comment]
> * Do we have infra capacity for the extra builds? [@jtmcdole](https://github.com/jtmcdole)

Infra for building mac engine and engine accessories for **every** pr

| Build | Orchestrator | Drones |Total|Running Tot|
| :--- | ---: | ---: | ---: | ---: |
| `Mac mac_ios_engine` | 1 | 9 | 11 | 11 |
| `Mac mac_ios_engine_ddm` | 1 | 4 | 5 | 16 |
| `Mac mac_host_engine` | 1 | 18 | 19 | 35 |
| `Linux mac_android_aot_engine` | 0 | 6  | 6 | 41 |

#191002 seeks to add 8 more to `mac_ios_engine`. At this moment, I would veto this addition of resources since nwe have only 134 drones; that PR would push our drones to 138, which is more than we have.

Even if  did content hashing to reduce the number of builds, I would still have reservations unless we had either more drones or faster builders - which we are working on, but that takes time.

**Comment by FelixMittermeier** [comment]
Just a hint: we could reduce it by just adding these new builds for Apple Silicon (so no support for Intel based simulators).
That would cut the additions in half I think but it would still be a huge benefit for most of the developers.

**Comment by jtmcdole** [comment]
> Just a hint: we could reduce it by just adding these new builds for Apple Silicon (so no support for Intel based simulators). That would cut the additions in half I think but it would still be a huge benefit for most of the developers.

I still wouldn't support doing this without clawing back more machines / builds. We have a few plans in the works for more mac machines.
